susemanager-utils/susemanager-sls/src/modules/mgr_caasp_manager.py

```diff
@@ -0,0 +1,347 @@
+                salt.utils.stringutils.to_str(skuba_proc.stderr))
+        log.error(error_msg)
+
+    ret = {
+        'stdout': salt.utils.stringutils.to_str(skuba_proc.stdout),
+        'stderr': salt.utils.stringutils.to_str(skuba_proc.stderr),
+        'success': not skuba_proc.process.returncode,
+        'retcode': skuba_proc.process.returncode,
+    }
+    return ret
+
+
+def add_node(skuba_cluster_path,
+             node_name,
+             role,
+             target,
+             ignore_preflight_errors=None,
+             port=None,
+             sudo=None,
+             user=None,
+             verbosity=None,
+             timeout=DEFAULT_TIMEOUT,
+             **kwargs):
+
+    cmd_args = "node join --role {} --target {} {}".format(role, target, node_name)
+
+    if ignore_preflight_errors:
+        cmd_args += " --ignore-preflight-errors {}".format(ignore_preflight_errors)
+    if port:
+        cmd_args += " --port {}".format(port)
+    if sudo:
+        cmd_args += " --sudo"
+    if user:
+        cmd_args += " --user {}".format(user)
+    if verbosity:
+        cmd_args += " --verbosity {}".format(verbosity)
+
+    skuba_proc = _call_skuba(skuba_cluster_path, cmd_args, timeout=timeout)
+    if skuba_proc.process.returncode != 0:
+        error_msg = "Unexpected error {} at skuba when adding a new node: {}".format(
+                skuba_proc.process.returncode,
+                salt.utils.stringutils.to_str(skuba_proc.stderr))
+        log.error(error_msg)
+
+    ret = {
+        'stdout': salt.utils.stringutils.to_str(skuba_proc.stdout),
+        'stderr': salt.utils.stringutils.to_str(skuba_proc.stderr),
+        'success': not skuba_proc.process.returncode,
+        'retcode': skuba_proc.process.returncode,
+    }
+    return ret
+
+
+def upgrade_cluster(skuba_cluster_path,
+                    verbosity=None,
+                    timeout=DEFAULT_TIMEOUT,
+                    **kwargs):
+
+    cmd_args = "cluster upgrade plan"
+
+    if verbosity:
+        cmd_args += " --verbosity {}".format(verbosity)
+
+    skuba_proc = _call_skuba(skuba_cluster_path, cmd_args, timeout=timeout)
+    if skuba_proc.process.returncode != 0:
+        error_msg = "Unexpected error {} at skuba when upgrading the cluster: {}".format(
+                skuba_proc.process.returncode,
+                salt.utils.stringutils.to_str(skuba_proc.stderr))
+        log.error(error_msg)
+
+    ret = {
+        'stdout': salt.utils.stringutils.to_str(skuba_proc.stdout),
+        'stderr': salt.utils.stringutils.to_str(skuba_proc.stderr),
+        'success': not skuba_proc.process.returncode,
+        'retcode': skuba_proc.process.returncode,
+    }
+    return ret
+
+
+def upgrade_node(skuba_cluster_path,
+                 verbosity=None,
+                 timeout=DEFAULT_TIMEOUT,
+                 plan=False,
+                 **kwargs):
+
+    cmd_args = "node upgrade {}".format("plan" if plan else "apply")
```

Comment:
Just a comment for the near future (non-blocking). When performing upgrades there are different stages when it comes to upgrading addons and nodes, so in general:

1. Upgrade addons to the latest version of the current platform version (`skuba addon upgrade plan/apply`)
2. Upgrade all nodes normally (`skuba node upgrade plan/apply`)
3. Upgrade addons to the new version of the platform (`skuba addon upgrade plan/apply`)

1. and 3. are not always required (depends on what changes on each release), but can always be run to ensure the most constraining scenario.

Reply:
@ereslibre thanks for clarifying the upgrade stages! I'll be pushing those missing methods on the next round of PR. :+1: